pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/queries/QueryTestCaseRunner.java

```diff
@@ -0,0 +1,126 @@
+        String tableName = e.getKey();
+        // TODO: able to choose table type, now default to OFFLINE
+        String tableNameWithType = TableNameBuilder.forType(TableType.OFFLINE).tableNameWithType(tableName);
+        org.apache.pinot.spi.data.Schema pinotSchema = constructSchema(tableName, e.getValue());
+        schemaMap.put(tableName, pinotSchema);
+        factory1.registerTable(pinotSchema, tableNameWithType);
+        factory2.registerTable(pinotSchema, tableNameWithType);
+      }
+      for (Map.Entry<String, List<Object[]>> e : testCase._tableInputs.entrySet()) {
+        String tableName = e.getKey();
+        String tableNameWithType = TableNameBuilder.forType(TableType.OFFLINE).tableNameWithType(tableName);
+        // TODO: able to select add rows to server1 or server2 (now default server1)
+        // TODO: able to select add rows to existing segment or create new one (now default create one segment)
+        factory1.addSegment(tableNameWithType, toRow(e.getValue()));
+      }
+      // add all the tables to
+      for (Map.Entry<String, Schema> entry: schemaMap.entrySet()) {
+        addTableToH2(entry.getValue(), Collections.singletonList(entry.getKey()));
+        addDataToH2(entry.getValue(), factory1.buildTableRowsMap());
+      }
+    }
+    QueryServerEnclosure server1 = new QueryServerEnclosure(factory1);
+    QueryServerEnclosure server2 = new QueryServerEnclosure(factory2);
+
+    _reducerGrpcPort = QueryTestUtils.getAvailablePort();
+    _reducerHostname = String.format("Broker_%s", QueryConfig.DEFAULT_QUERY_RUNNER_HOSTNAME);
+    Map<String, Object> reducerConfig = new HashMap<>();
+    reducerConfig.put(QueryConfig.KEY_OF_QUERY_RUNNER_PORT, _reducerGrpcPort);
+    reducerConfig.put(QueryConfig.KEY_OF_QUERY_RUNNER_HOSTNAME, _reducerHostname);
+    _mailboxService = new GrpcMailboxService(_reducerHostname, _reducerGrpcPort, new PinotConfiguration(reducerConfig));
+    _mailboxService.start();
+
+    _queryEnvironment = QueryEnvironmentTestBase.getQueryEnvironment(_reducerGrpcPort, server1.getPort(),
+        server2.getPort(), factory1.buildSchemaMap(), factory1.buildTableSegmentNameMap(),
+        factory2.buildTableSegmentNameMap());
+    server1.start();
+    server2.start();
+    // this doesn't test the QueryServer functionality so the server port can be the same as the mailbox port.
+    // this is only use for test identifier purpose.
+    int port1 = server1.getPort();
+    int port2 = server2.getPort();
+    _servers.put(new WorkerInstance("localhost", port1, port1, port1, port1), server1);
+    _servers.put(new WorkerInstance("localhost", port2, port2, port2, port2), server2);
+  }
+
+  @AfterClass
+  public void tearDown() {
+    DataTableBuilderFactory.setDataTableVersion(DataTableBuilderFactory.DEFAULT_VERSION);
+    for (QueryServerEnclosure server : _servers.values()) {
+      server.shutDown();
+    }
+    _mailboxService.shutdown();
+  }
+
+  @Test
```

Comment:
can we use composition instead of inheritance? instead of subclassing this test, we can have a util method `runTest(TestCase case)` and then the individual tests would just call that?